### PR TITLE
[AKS] `az aks create, aks nodepool add`: Add `--kubelet-config` and `--linux-os-config` to support custom node configuration

### PR DESCRIPTION
**Pull request by alexeldeib**
**Description**
Add custom node config for AKS

**Testing Guide**
Added live test example

**History Notes**

[AKS] `az aks create`: Add `--kubelet-config` and `--linux-os-config` to support custom node configuration
[AKS] `az aks nodepool add`: Add `--kubelet-config` and `--linux-os-config` to support custom node configuration

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

**Comment by alexeldeib** [comment]
@FumingZhang mind taking a look?

File: src/azure-cli/azure/cli/command_modules/acs/_params.py [hunk 1]
```diff
@@ -314,6 +314,8 @@ def load_arguments(self, _):
         c.argument('enable_sgxquotehelper', action='store_true')
         c.argument('enable_fips_image', action='store_true')
         c.argument('snapshot_id', validator=validate_snapshot_id)
+        c.argument('kubelet_config', type=str)
+        c.argument('linux_os_config', type=str)
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/acs/_params.py:318]: type=str is the default setting, so we don't need an explicit declaration

File: src/azure-cli/azure/cli/command_modules/acs/_params.py [hunk 2]
```diff
@@ -472,6 +474,8 @@ def load_arguments(self, _):
                        '--enable-ultra-ssd'], action='store_true')
             c.argument('enable_fips_image', action='store_true')
             c.argument('snapshot_id', validator=validate_snapshot_id)
+            c.argument('kubelet_config', type=str)
+            c.argument('linux_os_config', type=str)
```
Comment (wangzelin007) [inline src/azure-cli/azure/cli/command_modules/acs/_params.py:478]: same as above

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 3]
```diff
@@ -4238,3 +4251,110 @@ def aks_snapshot_list(cmd, client, resource_group_name=None):  # pylint: disable
...
+        "allowedUnsafeSysctls", None)
+    config_object.fail_swap_on = kubelet_config.get("failSwapOn", None)
+    config_object.container_log_max_files = kubelet_config.get(
+        "containerLogMaxFiles", None)
+    config_object.container_log_max_size_mb = kubelet_config.get(
+        "containerLogMaxSizeMB", None)
+    config_object.pod_max_pids = kubelet_config.get(
+        "podMaxPids", None)
+
+    return config_object
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:4285]: Actually, there is no need to convert a piece of json formatted content to our model. The serializer will convert it back to json format later.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 4]
```diff
@@ -4238,3 +4251,110 @@ def aks_snapshot_list(cmd, client, resource_group_name=None):  # pylint: disable
...
+    config_object.sysctls.fs_aio_max_nr = sysctls.get("fsAioMaxNr", None)
+    config_object.sysctls.fs_nr_open = sysctls.get("fsNrOpen", None)
+    config_object.sysctls.kernel_threads_max = sysctls.get(
+        "kernelThreadsMax", None)
+    config_object.sysctls.vm_max_map_count = sysctls.get("vmMaxMapCount", None)
+    config_object.sysctls.vm_swappiness = sysctls.get("vmSwappiness", None)
+    config_object.sysctls.vm_vfs_cache_pressure = sysctls.get(
+        "vmVfsCachePressure", None)
+
+    return config_object
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:4360]: Similar to the above.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py (hunk 3 above)
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:4285]:
```suggestion
    if not os.path.isfile(file_path):
        raise InvalidArgumentValueError("{} is not valid file, or not accessable.".format(file_path))
    kubelet_config = get_file_json(file_path)
    if not isinstance(kubelet_config, dict):
        msg = "Error reading kubelet configuration at {}. Please see https://aka.ms/CustomNodeConfig for proper format."
        raise InvalidArgumentValueError(msg.format(file_path))
    return kubelet_config
```

Please also use more specific error type.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py (hunk 4 above)
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:4360]:
```suggestion
    if not os.path.isfile(file_path):
        raise InvalidArgumentValueError("{} is not valid file, or not accessable.".format(file_path))
    os_config = get_file_json(file_path)
    if not isinstance(os_config, dict):
        msg = "Error reading Linux OS configuration at {}. \
            Please see https://aka.ms/CustomNodeConfig for proper format."
        raise InvalidArgumentValueError(msg.format(file_path))
    return os_config
```

File: src/azure-cli/azure/cli/command_modules/acs/_help.py [hunk 5]
```diff
@@ -503,6 +503,12 @@
   - name: --snapshot-id
     type: string
     short-summary: The source snapshot id used to create this cluster.
+  - name: --kubelet-config
+    type: string
+    short-summary: Kubelet configurations for agent nodes.
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/_help.py:508]: I am not sure if it's better to remind the user that this should be the path to a json format configuration file.

Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/_help.py:508]: Also a link indicates what properties/options the configuration file can have?

**Comment by alexeldeib** [comment]
hmm, not sure about that last one? it's not in my code, but the error is too many local vars, yet it's on a function definition? https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/serviceconnector/custom.py#L244

**Comment by FumingZhang** [comment]
The CI failure is fixed by PR [#21826](https://github.com/Azure/azure-cli/pull/21826).

**Review by FumingZhang — APPROVED** [review]
LGTM

**Review by wangzelin007 — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**